bundler/bundler.go

```diff
@@ -50,49 +73,144 @@ func BundleBytes(bytes []byte, configuration *datamodel.DocumentConfiguration) (
 // document will be a valid OpenAPI specification, containing no references.
 //
 // Circular references will not be resolved and will be skipped.
-func BundleDocument(model *v3.Document) ([]byte, error) {
-	return bundle(model, false)
+func BundleDocument(model *highV3.Document) ([]byte, error) {
+	return bundle(model, BundleOptions{RelativeRefHandling: RefHandlingInline})
 }
 
-func bundle(model *v3.Document, inline bool) ([]byte, error) {
+func bundle(model *highV3.Document, opts BundleOptions) (_ []byte, err error) {
 	rolodex := model.Rolodex
 
-	compact := func(idx *index.SpecIndex, root bool) {
-		mappedReferences := idx.GetMappedReferences()
-		sequencedReferences := idx.GetRawReferencesSequenced()
-		for _, sequenced := range sequencedReferences {
-			mappedReference := mappedReferences[sequenced.FullDefinition]
-
-			// if we're in the root document, don't bundle anything.
-			refExp := strings.Split(sequenced.FullDefinition, "#/")
-			if len(refExp) == 2 {
-				if refExp[0] == sequenced.Index.GetSpecAbsolutePath() || refExp[0] == "" {
-					if root && !inline {
-						idx.GetLogger().Debug("[bundler] skipping local root reference",
-							"ref", sequenced.Definition)
-						continue
-					}
-				}
+	idx := rolodex.GetRootIndex()
+	mappedReferences := idx.GetMappedReferences()
+	sequencedReferences := idx.GetRawReferencesSequenced()
+
+	for _, sequenced := range sequencedReferences {
+		mappedReference := mappedReferences[sequenced.FullDefinition]
+		if mappedReference == nil {
+			return nil, fmt.Errorf("no mapped reference found for: %s", sequenced.FullDefinition)
+		}
+
+		if mappedReference.DefinitionFile() == idx.GetSpecAbsolutePath() {
+			// Don't bundle anything that's in the main file.
+			continue
+		}
+
+		switch opts.RelativeRefHandling {
+		case RefHandlingInline:
+			// Just deal with simple inlining.
+			sequenced.Node.Content = mappedReference.Node.Content
+		case RefHandlingCompose:
+			// Recursively collect all reference targets to be bundled into the root
+			// file.
+			bundledComponents := make(map[string]*index.ReferenceNode)
+			if err := bundleRefTarget(sequenced, mappedReference, bundledComponents, opts); err != nil {
+				return nil, err
```

Comment:
Made a tiny bit of progress today. The idea is that `bundledComponents` now contains all the `yaml.Node`s which are to be bundled into the root spec, using their `Definition` (i.e. `#/components/schemas/MySchema`) as keys.

What I'm stuck with now, is that I don't know how to actually use these nodes. I figured the "right" way, you mentioned in your previous comments, would be to manipulate the `low` model by adding to the components there. I just can't make sense of the types used there.  See my next comment.